### ISSUE TITLE
error committing: invalid mutable ref: failed to mount cache

### ISSUE THREAD

**Issue by Anton-4** [earthly/earthly#1004]
The following command used to fail very sporadically for us, but it is now failing fairly often:
```
RUN --mount=type=cache,target=$SCCACHE_DIR cargo test --release && sccache --show-stats
```
This errors with the output:
```
================================ FAILURE [main] ================================
          +test-rust *failed* | --> RUN cargo test --release && sccache --show-stats
          +test-rust *failed* | [no output]
          +test-rust *failed* | ERROR: Command exited with non-zero code: RUN cargo test --release && sccache --show-stats
Error: error committing w693d003zzz4njaka75wqhhzn: invalid mutable ref 0xc0167f50e0: invalid: executor failed running [/bin/sh -c  /usr/bin/earth_debugger /bin/sh -c 'cargo test --release && sccache --show-stats']: failed to mount /earthbuild/sccache_dir: failed to mount w693d003zzz4njaka75wqhhzn: snapshot w693d003zzz4njaka75wqhhzn does not exist: not found
```

This [buildkit issue](https://github.com/moby/buildkit/issues/1956) seems to be the cause.

I have no experience with go so trying to fix the buildkit issue would be very time-consuming.
Instead, I'd like to fork earthly to log this error but continue execution.
This would allow my colleagues to use CI without requiring my help every time this issue occurs.

Can any of you tell me where in the codebase I can add the code to log the error but continue execution when this mounting failure occurs?

**Comment by Anton-4** [comment]
Here is the bash script for anyone sharing the same faith:
```bash
#!/usr/bin/env bash

LOG_FILE="earthly_log.txt"
touch $LOG_FILE

script -efq $LOG_FILE -c "earthly --config ci/earthly-conf.yml +test-all"
EXIT_CODE=$?

if grep -q "failed to mount" "$LOG_FILE"; then
  echo ""
  echo ""
  echo "------<<<<<<!!!!!!>>>>>>------"
  echo "DETECTED FAILURE TO MOUNT ERROR: running without cache"
  echo "------<<<<<<!!!!!!>>>>>>------"
  echo ""
  echo ""
  earthly --config ci/earthly-conf.yml --no-cache +test-all
else
  exit $EXIT_CODE
fi
```

**Comment by vladaionescu** [comment]
I wonder if increasing the cache size makes this occurrence rarer for you, as another workaround. Perhaps when the cache GC kicks in and clears this cache mount, it causes an issue with the next run not being able to use it. Larger cache would mean that this happens more rarely. https://docs.earthly.dev/docs/earthly-config#cache_size_mb

**Comment by Anton-4** [comment]
Thanks for your help! Yes, I was also thinking the cache size could be involved here. I have increased the cache size to 25 GB. Is there any way to log the amount of free space on the cache or log when the cache is being cleared?

**Comment by Anton-4** [comment]
I suppose ideally the cache should never be cleared midrun. Perhaps earthly/buildkit should work with a soft and a hard cache size limit. On exceeding the soft limit, the cache is cleared after the last command is run. On exceeding the hard limit it is cleared more instantly.

**Comment by vladaionescu** [comment]
Yeah, normally it doesn't clean up so aggressively. I think that the buildkit issue you referenced may be the root cause here.
